File: kie-wb-common-services/kie-wb-common-verifier/kie-wb-common-verifier-reporting-client/src/main/java/org/kie/workbench/common/services/verifier/reporting/client/controller/Events.java
/*
 * Copyright 2018 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.kie.workbench.common.services.verifier.reporting.client.controller;

import java.util.HashSet;
import java.util.Set;

import com.google.gwt.event.shared.EventBus;
import com.google.gwt.event.shared.HandlerRegistration;
import org.kie.soup.commons.validation.PortablePreconditions;
import org.kie.workbench.common.widgets.decoratedgrid.client.widget.events.AppendRowEvent;
import org.kie.workbench.common.widgets.decoratedgrid.client.widget.events.DeleteRowEvent;
import org.kie.workbench.common.widgets.decoratedgrid.client.widget.events.InsertRowEvent;
import org.kie.workbench.common.widgets.decoratedgrid.client.widget.events.UpdateColumnDataEvent;

public class Events {

    private final EventBus eventBus;
    private final AnalyzerControllerImpl analyzerController;

    private Set<HandlerRegistration> eventBusHandlerRegistrations = new HashSet<>();

    public Events(final EventBus eventBus,
                  final AnalyzerControllerImpl analyzerController) {
        this.eventBus = PortablePreconditions.checkNotNull("eventBus",
                                                           eventBus);
        this.analyzerController = PortablePreconditions.checkNotNull("analyzerController",
                                                                     analyzerController);
    }

    public void setup() {
        if (eventBusHandlerRegistrations.isEmpty()) {
            eventBusHandlerRegistrations.add(eventBus.addHandler(ValidateEvent.TYPE,
                                                                 analyzerController));
            eventBusHandlerRegistrations.add(eventBus.addHandler(DeleteRowEvent.TYPE,
                                                                 analyzerController));
            eventBusHandlerRegistrations.add(eventBus.addHandler(AfterColumnDeleted.TYPE,
                                                                 analyzerController));
            eventBusHandlerRegistrations.add(eventBus.addHandler(UpdateColumnDataEvent.TYPE,
                                                                 analyzerController));
            eventBusHandlerRegistrations.add(eventBus.addHandler(AppendRowEvent.TYPE,
                                                                 analyzerController));
            eventBusHandlerRegistrations.add(eventBus.addHandler(InsertRowEvent.TYPE,
                                                                 analyzerController));
            eventBusHandlerRegistrations.add(eventBus.addHandler(AfterColumnInserted.TYPE,
                                                                 analyzerController));
        }
    }

    public void teardown() {

        for (final HandlerRegistration handlerRegistration : eventBusHandlerRegistrations) {
            handlerRegistration.removeHandler();
        }

        eventBusHandlerRegistrations.clear();
    }
}

File: packages/sonataflow-operator/utils/events.go
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package utils

import (
	"context"
	"fmt"
	"net/http"

	cloudevents "github.com/cloudevents/sdk-go/v2"
	cehttp "github.com/cloudevents/sdk-go/v2/protocol/http"
)

// SendCloudEvent Sends a cloud event to the given url using the http protocol binding. By default, events are sent in
// binary mode.
func SendCloudEvent(event *cloudevents.Event, url string) error {
	return SendCloudEventWithContext(event, context.TODO(), url)
}

// SendCloudEventWithContext Sends a cloud event to the given url using the http protocol binding. By default, events
// are sent in binary mode.
func SendCloudEventWithContext(event *cloudevents.Event, ctx context.Context, url string) error {
	targetCtx := cloudevents.ContextWithTarget(ctx, url)
	p, err := cloudevents.NewHTTP()
	if err != nil {
		return err
	}
	c, err := cloudevents.NewClient(p, cloudevents.WithTimeNow(), cloudevents.WithUUIDs())
	if err != nil {
		return err
	}
	res := c.Send(targetCtx, *event)
	if cloudevents.IsUndelivered(res) {
		return fmt.Errorf("failed to send cloud event to url: %s, err: %s", url, res.Error())
	} else {
		var httpResult *cehttp.Result
		if cloudevents.ResultAs(res, &httpResult) {
			if !resultOK(httpResult) {
				return fmt.Errorf("failed to send cloud event to url: %s, err: %s", url, httpResult.Error())
			}
		} else {
			return fmt.Errorf("failed to send cloud event to url: %s, Send did not return an HTTP response: %s", url, res)
		}
	}
	return nil
}

func resultOK(httpResult *cehttp.Result) bool {
	return httpResult.StatusCode == http.StatusOK || httpResult.StatusCode == http.StatusAccepted
}
